Sagemaker Endpoint LLM

3coins opened the pull request:
Updates #965

nimimeht committed:
Adding a SagemakerEndpoint class (#953)

Commit:
cr

nimimeht committed:
Updates to Sagemaker Endpoint  (#1217)

3coins committed:
Added a content handler, updated notebook

jainsid24 commented:
@3coins how to provide a model name in this? Say I create an Sagemaker endpoint config with an LLM, how should I add that parameter when calling SagemakerEndpoint to chose the model?

jainsid24 commented on docs/modules/document_loaders/examples/sagemaker.ipynb:
```diff
@@ -0,0 +1,119 @@
+   "source": [
+    "from typing import Dict\n",
+    "\n",
+    "from langchain import PromptTemplate, SagemakerEndpoint, ContentHandlerBase\n",
```
ContentHandlerBase can't be import this way, this works ⬇️ 
```from langchain.llms.sagemaker_endpoint import ContentHandlerBase```

3coins replied:
@jainsid24 
Thanks for pointing out these points. Let me spend some more time today on this, will post an update soon.

3coins committed:
Fixed import

3coins committed:
Updated docs

3coins committed:
Added optional endpoint args, corrected notebook, fixed linting errors

3coins commented:
@jainsid24 
Fixed references in the notebook, also added a new optional dict as input where `TargetModel` and other arguments that `invoke_endpoint` supports can be passed. Thanks for reviewing this change.

jainsid24 commented:
@3coins thanks for adding the changes. This is now working as expected. I've been trying to use this on [Flan-T5-XXL](https://huggingface.co/google/flan-t5-xxl) model and it works flawlessly. You can add this optional section to the notebook as well for those who are using `Flan-T5` model.

```
class ContentHandler(ContentHandlerBase):
    content_type = "application/json"
    accepts = "application/json"

    def transform_input(self, prompt: str, model_kwargs: Dict) -> bytes:
        input_data = {"text_inputs": [prompt], **model_kwargs}
        input_str = json.dumps(input_data)
        return input_str.encode('utf-8')
    
    def transform_output(self, output: bytes) -> str:
        response_json = json.loads(output.read().decode("utf-8"))
        return response_json["generated_texts"][0]

content_handler = ContentHandler()

parameters = {
  "early_stopping": True,
  "length_penalty": 2.0,
  "temperature": 0,
  "min_length": 10,
  "no_repeat_ngram_size": 3,
}

chain = load_qa_chain(
    llm=SagemakerEndpoint(
        endpoint_name="jumpstart-dft-hf-text2text-flan-t5-xl", 
        credentials_profile_name="default", 
        region_name="us-west-2", 
        model_kwargs=parameters,
        content_handler=content_handler
    ),
    prompt=PROMPT
)

chain({"input_documents": docs, "question": query}, return_only_outputs=True)
```

@hwchase17 We will use this for production usecases, have tested this. Would be great if we can fast-pace this PR :)

hwchase17 commented:
awesome! fast-tracked for release tmrw - thanks @3coins and @jainsid24 !

3coins commented:
@hwchase17 
Thanks for merging.
